File: artifacts/discord-bot/src/cogs/moderation.js
/**
 * MODERATION COG
 * Commands: ban, kick, mute, unmute, warn, warnings, purge, slowmode
 * Supports both prefix (!) and slash (/) commands
 */

import { SlashCommandBuilder, PermissionFlagsBits } from 'discord.js';
import { successEmbed, errorEmbed, warnEmbed, brandEmbed } from '../utils/embed.js';
import { db } from '../utils/database.js';

// ─── Database Setup ───────────────────────────────────────────────────────────

db.exec(`
  CREATE TABLE IF NOT EXISTS warnings (
    id        INTEGER PRIMARY KEY AUTOINCREMENT,
    guild_id  TEXT NOT NULL,
    user_id   TEXT NOT NULL,
    mod_id    TEXT NOT NULL,
    reason    TEXT NOT NULL,
    created_at INTEGER NOT NULL DEFAULT (strftime('%s', 'now'))
  );
`);

// ─── Helpers ──────────────────────────────────────────────────────────────────

function hasPerms(member, perm) {
  return member.permissions.has(perm);
}

// ─── Prefix Commands ──────────────────────────────────────────────────────────

const banCommand = {
  name: 'ban',
  description: 'Ban a member',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.BanMembers))
      return message.reply({ embeds: [errorEmbed('You need the **Ban Members** permission.')] });

    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null);
    if (!member) return message.reply({ embeds: [errorEmbed('Member not found.')] });
    if (!member.bannable) return message.reply({ embeds: [errorEmbed('I cannot ban that member.')] });

    const reason = args.slice(1).join(' ') || 'No reason provided';
    await member.ban({ reason });
    await message.reply({ embeds: [successEmbed(`Banned **${member.user.tag}** — ${reason}`)] });
  },
};

const kickCommand = {
  name: 'kick',
  description: 'Kick a member',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.KickMembers))
      return message.reply({ embeds: [errorEmbed('You need the **Kick Members** permission.')] });

    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null);
    if (!member) return message.reply({ embeds: [errorEmbed('Member not found.')] });
    if (!member.kickable) return message.reply({ embeds: [errorEmbed('I cannot kick that member.')] });

    const reason = args.slice(1).join(' ') || 'No reason provided';
    await member.kick(reason);
    await message.reply({ embeds: [successEmbed(`Kicked **${member.user.tag}** — ${reason}`)] });
  },
};

const muteCommand = {
  name: 'mute',
  aliases: ['timeout'],
  description: 'Timeout a member (in minutes)',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.ModerateMembers))
      return message.reply({ embeds: [errorEmbed('You need the **Moderate Members** permission.')] });

    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null);
    if (!member) return message.reply({ embeds: [errorEmbed('Member not found.')] });

    const minutes = parseInt(args[1]) || 10;
    const reason = args.slice(2).join(' ') || 'No reason provided';
    await member.timeout(minutes * 60 * 1000, reason);
    await message.reply({ embeds: [successEmbed(`Muted **${member.user.tag}** for **${minutes}m** — ${reason}`)] });
  },
};

const unmuteCommand = {
  name: 'unmute',
  description: 'Remove timeout from a member',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.ModerateMembers))
      return message.reply({ embeds: [errorEmbed('You need the **Moderate Members** permission.')] });

    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null);
    if (!member) return message.reply({ embeds: [errorEmbed('Member not found.')] });

    await member.timeout(null);
    await message.reply({ embeds: [successEmbed(`Unmuted **${member.user.tag}**`)] });
  },
};

const warnCommand = {
  name: 'warn',
  description: 'Warn a member',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.ModerateMembers))
      return message.reply({ embeds: [errorEmbed('You need the **Moderate Members** permission.')] });

    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null);
    if (!member) return message.reply({ embeds: [errorEmbed('Member not found.')] });

    const reason = args.slice(1).join(' ') || 'No reason provided';
    db.prepare('INSERT INTO warnings (guild_id, user_id, mod_id, reason) VALUES (?, ?, ?, ?)')
      .run(message.guild.id, member.id, message.author.id, reason);

    const count = db.prepare('SELECT COUNT(*) as c FROM warnings WHERE guild_id=? AND user_id=?')
      .get(message.guild.id, member.id).c;

    await message.reply({ embeds: [warnEmbed(`Warned **${member.user.tag}** — ${reason}\nTotal warnings: **${count}**`)] });

    await member.send(`⚠️ You have been warned in **${message.guild.name}**\nReason: ${reason}`).catch(() => {});
  },
};

const warningsCommand = {
  name: 'warnings',
  aliases: ['warns'],
  description: 'View warnings for a member',
  cog: 'Moderation',
  async execute(message, args) {
    const member = message.mentions.members.first()
      || await message.guild.members.fetch(args[0]).catch(() => null)
      || message.member;

    const rows = db.prepare('SELECT * FROM warnings WHERE guild_id=? AND user_id=? ORDER BY created_at DESC LIMIT 10')
      .all(message.guild.id, member.id);

    const embed = brandEmbed(`⚠️ Warnings for ${member.user.tag}`)
      .setDescription(rows.length
        ? rows.map((w, i) => `**${i + 1}.** ${w.reason} — <@${w.mod_id}> <t:${w.created_at}:R>`).join('\n')
        : 'No warnings found.');

    await message.reply({ embeds: [embed] });
  },
};

const purgeCommand = {
  name: 'purge',
  aliases: ['clear', 'clean'],
  description: 'Bulk delete messages',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.ManageMessages))
      return message.reply({ embeds: [errorEmbed('You need the **Manage Messages** permission.')] });

    const amount = Math.min(parseInt(args[0]) || 10, 100);
    await message.channel.bulkDelete(amount + 1, true).catch(() => {});
    const confirm = await message.channel.send({ embeds: [successEmbed(`Deleted **${amount}** message(s).`)] });
    setTimeout(() => confirm.delete().catch(() => {}), 4000);
  },
};

const slowmodeCommand = {
  name: 'slowmode',
  description: 'Set channel slowmode (seconds)',
  cog: 'Moderation',
  async execute(message, args) {
    if (!hasPerms(message.member, PermissionFlagsBits.ManageChannels))
      return message.reply({ embeds: [errorEmbed('You need the **Manage Channels** permission.')] });

    const seconds = parseInt(args[0]) ?? 0;
    await message.channel.setRateLimitPerUser(seconds);
    await message.reply({ embeds: [successEmbed(`Slowmode set to **${seconds}s**`)] });
  },
};

// ─── Slash Commands ───────────────────────────────────────────────────────────

const slashBan = {
  data: new SlashCommandBuilder()
    .setName('ban')
    .setDescription('Ban a member')
    .setDefaultMemberPermissions(PermissionFlagsBits.BanMembers)
    .addUserOption((o) => o.setName('user').setDescription('User to ban').setRequired(true))
    .addStringOption((o) => o.setName('reason').setDescription('Reason')),

  async execute(interaction) {
    const member = interaction.options.getMember('user');
    const reason = interaction.options.getString('reason') ?? 'No reason provided';
    if (!member?.bannable)
      return interaction.reply({ embeds: [errorEmbed('Cannot ban that member.')], ephemeral: true });
    await member.ban({ reason });
    await interaction.reply({ embeds: [successEmbed(`Banned **${member.user.tag}** — ${reason}`)] });
  },
};

const slashKick = {
  data: new SlashCommandBuilder()
    .setName('kick')
    .setDescription('Kick a member')
    .setDefaultMemberPermissions(PermissionFlagsBits.KickMembers)
    .addUserOption((o) => o.setName('user').setDescription('User to kick').setRequired(true))
    .addStringOption((o) => o.setName('reason').setDescription('Reason')),

  async execute(interaction) {
    const member = interaction.options.getMember('user');
    const reason = interaction.options.getString('reason') ?? 'No reason provided';
    if (!member?.kickable)
      return interaction.reply({ embeds: [errorEmbed('Cannot kick that member.')], ephemeral: true });
    await member.kick(reason);
    await interaction.reply({ embeds: [successEmbed(`Kicked **${member.user.tag}** — ${reason}`)] });
  },
};

const slashMute = {
  data: new SlashCommandBuilder()
    .setName('mute')
    .setDescription('Timeout a member')
    .setDefaultMemberPermissions(PermissionFlagsBits.ModerateMembers)
    .addUserOption((o) => o.setName('user').setDescription('User to mute').setRequired(true))
    .addIntegerOption((o) => o.setName('minutes').setDescription('Duration in minutes').setRequired(true))
    .addStringOption((o) => o.setName('reason').setDescription('Reason')),

  async execute(interaction) {
    const member = interaction.options.getMember('user');
    const minutes = interaction.options.getInteger('minutes');
    const reason = interaction.options.getString('reason') ?? 'No reason provided';
    await member.timeout(minutes * 60 * 1000, reason);
    await interaction.reply({ embeds: [successEmbed(`Muted **${member.user.tag}** for **${minutes}m** — ${reason}`)] });
  },
};

const slashWarn = {
  data: new SlashCommandBuilder()
    .setName('warn')
    .setDescription('Warn a member')
    .setDefaultMemberPermissions(PermissionFlagsBits.ModerateMembers)
    .addUserOption((o) => o.setName('user').setDescription('User to warn').setRequired(true))
    .addStringOption((o) => o.setName('reason').setDescription('Reason')),

  async execute(interaction) {
    const member = interaction.options.getMember('user');
    const reason = interaction.options.getString('reason') ?? 'No reason provided';
    db.prepare('INSERT INTO warnings (guild_id, user_id, mod_id, reason) VALUES (?, ?, ?, ?)')
      .run(interaction.guild.id, member.id, interaction.user.id, reason);
    const count = db.prepare('SELECT COUNT(*) as c FROM warnings WHERE guild_id=? AND user_id=?')
      .get(interaction.guild.id, member.id).c;
    await interaction.reply({ embeds: [warnEmbed(`Warned **${member.user.tag}** — ${reason}\nTotal warnings: **${count}**`)] });
    await member.send(`⚠️ You have been warned in **${interaction.guild.name}**\nReason: ${reason}`).catch(() => {});
  },
};

const slashPurge = {
  data: new SlashCommandBuilder()
    .setName('purge')
    .setDescription('Bulk delete messages')
    .setDefaultMemberPermissions(PermissionFlagsBits.ManageMessages)
    .addIntegerOption((o) => o.setName('amount').setDescription('Number of messages (1-100)').setRequired(true)),

  async execute(interaction) {
    const amount = Math.min(interaction.options.getInteger('amount'), 100);
    await interaction.deferReply({ ephemeral: true });
    await interaction.channel.bulkDelete(amount, true);
    await interaction.editReply({ embeds: [successEmbed(`Deleted **${amount}** message(s).`)] });
  },
};

// ─── Setup ────────────────────────────────────────────────────────────────────

export default {
  setup(client) {
    const prefixCmds = [banCommand, kickCommand, muteCommand, unmuteCommand, warnCommand, warningsCommand, purgeCommand, slowmodeCommand];
    const slashCmds = [slashBan, slashKick, slashMute, slashWarn, slashPurge];

    for (const cmd of prefixCmds) client.commands.set(cmd.name, cmd);
    for (const cmd of slashCmds) client.slashCommands.set(cmd.data.name, cmd);
  },
};

File: artifacts/discord-bot/modules/moderation.py
"""
MODERATION MODULE
Commands: ban, kick, mute, unmute, warn, warnings, clearwarns, purge, slowmode, lock, unlock
Both prefix and slash variants.
"""
import discord
from discord import app_commands
from discord.ext import commands

from main.utils.database import add_warning, get_warnings, clear_warnings
from main.utils.embeds import success, error, warn, brand


class Moderation(commands.Cog, name="Moderation"):
    """🔨 Moderation commands."""

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    # ── !ban ─────────────────────────────────────────────────────────────────

    @commands.command(name="ban")
    @commands.has_permissions(ban_members=True)
    @commands.bot_has_permissions(ban_members=True)
    @commands.guild_only()
    async def ban_cmd(self, ctx: commands.Context, member: discord.Member, *, reason: str = "No reason provided"):
        """Ban a member from the server."""
        if member.top_role >= ctx.author.top_role and ctx.author.id != ctx.guild.owner_id:
            return await ctx.reply(embed=error("You can't ban someone with an equal or higher role."), mention_author=False)
        await member.ban(reason=f"{ctx.author} — {reason}")
        await ctx.reply(embed=success(f"Banned **{member}**\nReason: {reason}"), mention_author=False)
        try:
            await member.send(embed=error(f"You were **banned** from **{ctx.guild.name}**.\nReason: {reason}"))
        except discord.Forbidden:
            pass

    # ── !kick ─────────────────────────────────────────────────────────────────

    @commands.command(name="kick")
    @commands.has_permissions(kick_members=True)
    @commands.bot_has_permissions(kick_members=True)
    @commands.guild_only()
    async def kick_cmd(self, ctx: commands.Context, member: discord.Member, *, reason: str = "No reason provided"):
        """Kick a member from the server."""
        if member.top_role >= ctx.author.top_role and ctx.author.id != ctx.guild.owner_id:
            return await ctx.reply(embed=error("You can't kick someone with an equal or higher role."), mention_author=False)
        await member.kick(reason=f"{ctx.author} — {reason}")
        await ctx.reply(embed=success(f"Kicked **{member}**\nReason: {reason}"), mention_author=False)
        try:
            await member.send(embed=warn(f"You were **kicked** from **{ctx.guild.name}**.\nReason: {reason}"))
        except discord.Forbidden:
            pass

    # ── !mute ─────────────────────────────────────────────────────────────────

    @commands.command(name="mute", aliases=["timeout"])
    @commands.has_permissions(moderate_members=True)
    @commands.bot_has_permissions(moderate_members=True)
    @commands.guild_only()
    async def mute_cmd(self, ctx: commands.Context, member: discord.Member, minutes: int = 10, *, reason: str = "No reason provided"):
        """Timeout (mute) a member for X minutes."""
        import datetime
        duration = datetime.timedelta(minutes=minutes)
        await member.timeout(duration, reason=f"{ctx.author} — {reason}")
        await ctx.reply(embed=success(f"Muted **{member}** for **{minutes}m**\nReason: {reason}"), mention_author=False)

    # ── !unmute ───────────────────────────────────────────────────────────────

    @commands.command(name="unmute", aliases=["untimeout"])
    @commands.has_permissions(moderate_members=True)
    @commands.bot_has_permissions(moderate_members=True)
    @commands.guild_only()
    async def unmute_cmd(self, ctx: commands.Context, member: discord.Member):
        """Remove a timeout from a member."""
        await member.timeout(None)
        await ctx.reply(embed=success(f"Unmuted **{member}**"), mention_author=False)

    # ── !warn ─────────────────────────────────────────────────────────────────

    @commands.command(name="warn")
    @commands.has_permissions(moderate_members=True)
    @commands.guild_only()
    async def warn_cmd(self, ctx: commands.Context, member: discord.Member, *, reason: str = "No reason provided"):
        """Warn a member."""
        count = await add_warning(ctx.guild.id, member.id, ctx.author.id, reason)
        await ctx.reply(embed=warn(f"Warned **{member}**\nReason: {reason}\nTotal warnings: **{count}**"), mention_author=False)
        try:
            await member.send(embed=warn(f"You were **warned** in **{ctx.guild.name}**.\nReason: {reason}"))
        except discord.Forbidden:
            pass

    # ── !warnings ─────────────────────────────────────────────────────────────

    @commands.command(name="warnings", aliases=["warns"])
    @commands.guild_only()
    async def warnings_cmd(self, ctx: commands.Context, member: discord.Member = None):
        """View warnings for a member."""
        m = member or ctx.author
        rows = await get_warnings(ctx.guild.id, m.id)
        embed = brand(f"⚠️ Warnings for {m}")
        if rows:
            lines = [
                f"**{i+1}.** {r['reason']} — <@{r['mod_id']}> <t:{r['created_at']}:R>"
                for i, r in enumerate(rows)
            ]
            embed.description = "\n".join(lines)
        else:
            embed.description = "No warnings on record."
        await ctx.reply(embed=embed, mention_author=False)

    # ── !clearwarns ───────────────────────────────────────────────────────────

    @commands.command(name="clearwarns", aliases=["clearwarnings", "resetwarns"])
    @commands.has_permissions(moderate_members=True)
    @commands.guild_only()
    async def clearwarns_cmd(self, ctx: commands.Context, member: discord.Member):
        """Clear all warnings for a member."""
        count = await clear_warnings(ctx.guild.id, member.id)
        await ctx.reply(embed=success(f"Cleared **{count}** warning(s) for **{member}**."), mention_author=False)

    # ── !purge ────────────────────────────────────────────────────────────────

    @commands.command(name="purge", aliases=["clear", "clean"])
    @commands.has_permissions(manage_messages=True)
    @commands.bot_has_permissions(manage_messages=True)
    @commands.guild_only()
    async def purge_cmd(self, ctx: commands.Context, amount: int = 10):
        """Bulk delete messages (1–100)."""
        amount = max(1, min(amount, 100))
        await ctx.message.delete()
        deleted = await ctx.channel.purge(limit=amount)
        msg = await ctx.send(embed=success(f"Deleted **{len(deleted)}** message(s)."))
        import asyncio
        await asyncio.sleep(4)
        await msg.delete()

    # ── !slowmode ─────────────────────────────────────────────────────────────

    @commands.command(name="slowmode", aliases=["slow"])
    @commands.has_permissions(manage_channels=True)
    @commands.bot_has_permissions(manage_channels=True)
    @commands.guild_only()
    async def slowmode_cmd(self, ctx: commands.Context, seconds: int = 0):
        """Set channel slowmode in seconds (0 to disable)."""
        await ctx.channel.edit(slowmode_delay=max(0, min(seconds, 21600)))
        msg = f"Slowmode set to **{seconds}s**" if seconds else "Slowmode **disabled**"
        await ctx.reply(embed=success(msg), mention_author=False)

    # ── !lock / !unlock ───────────────────────────────────────────────────────

    @commands.command(name="lock")
    @commands.has_permissions(manage_channels=True)
    @commands.bot_has_permissions(manage_channels=True)
    @commands.guild_only()
    async def lock_cmd(self, ctx: commands.Context, channel: discord.TextChannel = None):
        """Lock a channel so members can't send messages."""
        ch = channel or ctx.channel
        await ch.set_permissions(ctx.guild.default_role, send_messages=False)
        await ctx.reply(embed=success(f"🔒 Locked {ch.mention}"), mention_author=False)

    @commands.command(name="unlock")
    @commands.has_permissions(manage_channels=True)
    @commands.bot_has_permissions(manage_channels=True)
    @commands.guild_only()
    async def unlock_cmd(self, ctx: commands.Context, channel: discord.TextChannel = None):
        """Unlock a channel."""
        ch = channel or ctx.channel
        await ch.set_permissions(ctx.guild.default_role, send_messages=None)
        await ctx.reply(embed=success(f"🔓 Unlocked {ch.mention}"), mention_author=False)

    # ── Slash Commands ────────────────────────────────────────────────────────

    @app_commands.command(name="ban", description="Ban a member")
    @app_commands.default_permissions(ban_members=True)
    @app_commands.guild_only()
    async def slash_ban(self, interaction: discord.Interaction, member: discord.Member, reason: str = "No reason provided"):
        if not member.guild_permissions.ban_members or not interaction.guild.me.guild_permissions.ban_members:
            return await interaction.response.send_message(embed=error("Missing permissions."), ephemeral=True)
        await member.ban(reason=reason)
        await interaction.response.send_message(embed=success(f"Banned **{member}** — {reason}"))

    @app_commands.command(name="kick", description="Kick a member")
    @app_commands.default_permissions(kick_members=True)
    @app_commands.guild_only()
    async def slash_kick(self, interaction: discord.Interaction, member: discord.Member, reason: str = "No reason provided"):
        await member.kick(reason=reason)
        await interaction.response.send_message(embed=success(f"Kicked **{member}** — {reason}"))

    @app_commands.command(name="mute", description="Timeout a member")
    @app_commands.default_permissions(moderate_members=True)
    @app_commands.guild_only()
    async def slash_mute(self, interaction: discord.Interaction, member: discord.Member, minutes: int = 10, reason: str = "No reason provided"):
        import datetime
        await member.timeout(datetime.timedelta(minutes=minutes), reason=reason)
        await interaction.response.send_message(embed=success(f"Muted **{member}** for **{minutes}m** — {reason}"))

    @app_commands.command(name="warn", description="Warn a member")
    @app_commands.default_permissions(moderate_members=True)
    @app_commands.guild_only()
    async def slash_warn(self, interaction: discord.Interaction, member: discord.Member, reason: str = "No reason provided"):
        count = await add_warning(interaction.guild.id, member.id, interaction.user.id, reason)
        await interaction.response.send_message(embed=warn(f"Warned **{member}** — {reason}\nTotal warnings: **{count}**"))

    @app_commands.command(name="purge", description="Bulk delete messages")
    @app_commands.default_permissions(manage_messages=True)
    @app_commands.guild_only()
    async def slash_purge(self, interaction: discord.Interaction, amount: int = 10):
        amount = max(1, min(amount, 100))
        await interaction.response.defer(ephemeral=True)
        deleted = await interaction.channel.purge(limit=amount)
        await interaction.followup.send(embed=success(f"Deleted **{len(deleted)}** message(s)."), ephemeral=True)


async def setup(bot: commands.Bot):
    await bot.add_cog(Moderation(bot))
